server/aesdsocket.c

```diff
@@ -1,196 +1,429 @@
-#include <netdb.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <sys/socket.h>
 #include <sys/types.h>
-#include <unistd.h>
+#include <sys/socket.h>
+#include <netdb.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
 #include <syslog.h>
+#include <fcntl.h>
+#include <unistd.h>
 #include <errno.h>
+#include <stdbool.h>
 #include <signal.h>
-#include <fcntl.h>
-#include <netinet/in.h>
-#include <arpa/inet.h>
+#include <pthread.h>
+#include <sys/time.h>
+#include <time.h>
+#include "queue.h"
 
-const static char *kPort = "9000";
+const static int kPort = 9000;
 const static char *kSocketData = "/var/tmp/aesdsocketdata";
-const static int kBufferSize = 1000000;
+const static int kBufferStartLength = 128;
 
 static int sfd = 0;
-static int cfd = 0;
 static int fd = 0;
 
-static void signal_handler(int signo)
+typedef struct socketThreadData
 {
-  syslog(LOG_INFO, "Caught signal, exiting.");
-  shutdown(sfd, SHUT_RDWR);
-  close(sfd);
-  close(fd);
-  unlink(kSocketData);
-  close(cfd);
-  closelog();
-  exit(0);
+  pthread_mutex_t *mutexExit;
```

Comment:
@qrogers6 some detail about what `mutexExit` is locking and maybe some more details about the other structure members would be good.  I'm not sure the mutex is necessary.  Sometimes it's held for writes of exit, sometimes for writes of complete, and sometimes for reads of exit.  If you only set exit in one thread and only read it outside the thread you shouldn't need locking.